File: src/main/java/com/viking/algorithms/search/S02_BinarySearch.java
package com.viking.algorithms.search;

/**
 * Author : Viking Den<dengwj@gionee.com>
 * Time : 2/1/18 3:01 PM
 */
public class S02_BinarySearch implements BaseSearch {

    @Override
    public void search() {
        int[] items = new int[]{2, 5, 8, 12, 16, 23, 38, 56, 72, 91} ;
        int   target = 23 ;
        binarySearch(items, target);

        int result = binarySearchByRecursive(items, target, 0, items.length - 1);
        if (result == -1){
            System.out.println("No binary search result with the given target!");
        }else{
            System.out.println("Binary search result index : " + result);
        }
    }

    private void binarySearch(int[] items, int target){
        int   result = -1 ;
        int   low = 0, high = items.length - 1 ;
        int   median ;
        while (low <= high){
            median = (low + high) / 2 ;
            if (items[median] == target){
                result = median ;
                break ;
            }
            if (items[median] > target){
                high = median - 1 ;
            }else{
                low = median + 1 ;
            }
        }
        if (result == -1){
            System.out.println("No binary search result with the given target.");
        }else{
            System.out.println("binary search result index : " + result);
        }
    }

    private int binarySearchByRecursive(int[] items, int target, int left, int right){
        if (left <= right){
            int median = ( left + right ) / 2 ;
            if (items[median] == target){
                return median ;
            }
            if (items[median] < target){
                return binarySearchByRecursive(items, target, median + 1, right) ;
            }
            return binarySearchByRecursive(items, target, left, median - 1) ;
        }else{
            return -1 ;
        }
    }

    public static void main(String[] args) {
        S02_BinarySearch binarySearch = new S02_BinarySearch() ;
        binarySearch.search();
    }
}

File: src/main/java/com/viking/algorithms/search/S02_BinarySearch.py
#!/usr/bin/python3
#coding=utf-8


def main():
    items = (2, 5, 8, 12, 16, 23, 38, 56, 72, 91)
    target = 23
    binarySearch(items, target)

    result = binarySearchByRecursive(items, target, 0, len(items) - 1)
    if result == -1:
        print('No binary search result with the given target.')
    else:
        print('Binary search result index : {}'.format(result))


def binarySearch(items, target):
    low, high = 0, len(items)-1
    result = -1
    median = -1
    while low <= high:
        median = (low + high) // 2
        if items[median] == target:
            result = median
            break
        if items[median] > target:
            high = median - 1
        else:
            low = median + 1
    if result == -1:
        print('No binary search result with the given target!')
    else:
        print('Binary search result index : {}'.format(result))


def binarySearchByRecursive(items, target, left, right):
    if left <= right:
        median = (left + right) // 2
        if items[median] == target:
            return median
        if items[median] > target:
            return binarySearchByRecursive(items, target, left, median - 1)
        return binarySearchByRecursive(items, target, median + 1, right)
    else:
        return -1


if __name__ == '__main__':
    main()
